### ISSUE TITLE
Add exceptions to the CPU and align LOAD/STORE/PUSH/POP

### ISSUE THREAD

**Issue by Zardoz89** [trillek-team/trillek-computer#27]
As the talk on the [IRC](http://logs.trillek.org/#id-262508) suggested, we should enforce LOAD/STORE like instructions and add exception interrupts to the CPU.

Aligned LOAD/STORE should be based on the size of the read/write, so : 
- LOADB/STOREB is always aligned
- LOADW/STOREW is aligned if address LSB == 0
- LOAD/STORE is aligned if address two LSB == 00
- PUSH/POP is aligned if address two LSB == 00

Independently of if we add exceptions and it are enabled, LOADW/STOREW/LOAD/STORE/PUSH/POP must ignore the appropriate LSB bits when reads/writes to the RAM.

Exception list : 
- Division error (division by zero)
- Overflow
- Unaligned LOAD/STORE
- Unaligned PUSH/POP (unaligned stack access)
- Illegal instruction

How enable/disable it and what interrupt message use, it's actually open. To enable/disable it ideally should be using the %flag register. For the message, we could keep the LSB to 0x00, so it's shared with the single step mode or we could reserve messages with LSB 0x00 for traps, and with LSB 0x01 for exceptions, using on the MSB of the message to indicate what exception happen (but this last would be redundant with the %flag register function)

**Comment by milesrout** [comment]
Has this been resolved?

**Comment by Zardoz89** [comment]
Nope. I think that this need a more clear and straightforward rethink.
The previous commit only make more clear some stuff related to this, but not changed behaviour or functionality beyond of fixing Single-Step mode.

**Comment by Zardoz89** [comment]
Trying to simply all the stuff and enforce align. I propose this : 
- LOADB/STOREB is always aligned to byte boundary.
- LOADW/STOREW is always aligned to word boundary. Ignores the LSB bit.
- LOAD/STORE is always aligned to double word boundary. Ignores the two LSB bits.
- %sp two LSB bits are always zero. PUSH/POP never could be unaligned.
- %ia two LSB bits are always zero, to avoid that an Interrupt handler become unaligned.

This this, we avoid adding unaligned exceptions, and keeps simple the CPU.

Also, we could add Invalid instruction exception.

**Comment by milesrout** [comment]
We should not ignore anything. If the LSB bits are aligned then we should fail fast and tell the program that they are illegal.
